src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php

```diff
@@ -226,20 +225,37 @@ protected function renderView(string $view, array $parameters = []): string
             throw new \LogicException('You cannot use the "renderView" method if the Twig Bundle is not available. Try running "composer require symfony/twig-bundle".');
         }
 
+        foreach ($parameters as $k => $v) {
+            if ($v instanceof FormInterface) {
+                $parameters[$k] = $v->createView();
+            }
+        }
+
         return $this->container->get('twig')->render($view, $parameters);
     }
 
     /**
      * Renders a view.
+     *
+     * If an invalid form is found in the list of parameters, a 422 status code is returned.
      */
     protected function render(string $view, array $parameters = [], Response $response = null): Response
     {
-        $content = $this->renderView($view, $parameters);
-
         if (null === $response) {
             $response = new Response();
         }
 
+        if (200 === $response->getStatusCode()) {
+            foreach ($parameters as $v) {
+                if ($v instanceof FormInterface && $v->isSubmitted() && !$v->isValid()) {
```

Comment:
When a `FormView` is passed (classic use case) this feature is not triggered, what about adding something like:
```php
|| ($v instanceof FormView && !($v->vars['valid'] ?? true))
```
or with an `elseif`?

Reply:
This could be done, but this would change the behavior of existing apps, which would start returning 421 instead of 200 currently. Not sure we want to do that.

Reply:
Returning a 200 with an invalid form is a bug from an HTTP cache perspective, and does not make sense if the method is not idempotent.
Maybe we want to fix this as part of such feature?

Reply:
We could. WDYT @symfony/mergers?

Reply:
@nicolas-grekas you're right, changing the behavior might break some test suites.

I've got an idea of a BC layer to opt-in and change the default in 7.0. Maybe I can send a PR later on to discuss it?

Reply:
feel free to follow up of course

Reply:
For the record, see #47600